### PR TITLE
[Debugger] Introduce log points support

### PR DESCRIPTION
**Pull request by NipunaRanasinghe**
## Purpose
This PR introduces logpoint support for Ballerina debugging in VSCode. Logpoints allows to inject logging into running programs without interfering with the normal functionality and, every time the program executes code at the logpoint location, Debugger logs a message in the console instead of suspending the remote VM.

This PR also adds debugger integration test scenarios related to the aforementioned feature and fixes https://github.com/ballerina-platform/ballerina-lang/issues/31535

## Samples
ezgif com-gif-maker

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [x] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-debugger-integration-test/src/main/java/org/ballerinalang/debugger/test/utils/DebugOutputListener.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.TimerTask;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/**
+ * Timer Task implementation to capture debugger output from server output events.
+ */
+public class DebugOutputListener extends TimerTask {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DebugOutputListener.class);
+    TestDAPClientConnector connector;
```
Comment (prabushi) [inline tests/jballerina-debugger-integration-test/src/main/java/org/ballerinalang/debugger/test/utils/DebugOutputListener.java:35]: Shall we make the 'connector' private as it has a public getter?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java [hunk 2]
```diff
@@ -176,6 +146,69 @@ private void processEvent(EventSet eventSet, Event event) {
...
+                    notifyStopEvent(event);
+                }
+            } else {
+                // As we are disabling all the breakpoint requests before evaluating the user's conditional expression,
+                // need to re-enable all the breakpoints before continuing the remote VM execution.
+                restoreBreakpoints(context.getLastInstruction());
+                context.getDebuggeeVM().resume();
+            }
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            context.getAdapter().sendOutput("Warning: Skipping the conditional breakpoint at line: "
```
Comment (prabushi) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java:200]: Shouldn't we consider the 'e' exception's details for the output?

File: tests/jballerina-debugger-integration-test/src/main/java/org/ballerinalang/debugger/test/utils/DebugOutputListener.java (hunk 1 above)
Comment (NipunaRanasinghe) [reply tests/jballerina-debugger-integration-test/src/main/java/org/ballerinalang/debugger/test/utils/DebugOutputListener.java]: Made private in c15d1af

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java (hunk 2 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java]: @prabushi As we show this warning in the debug console during the debug session, I doubt whether we need to expose java level exception details to the user.

Comment (prabushi) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java]: Maybe we can add them to debug level logs later.

**Review by prabushi — DISMISSED** [review]

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java [hunk 3]
```diff
@@ -176,6 +147,69 @@ private void processEvent(EventSet eventSet, Event event) {
...
+                breakpoint.getCondition().get() : "";
+        String logMessage = breakpoint.getLogMessage().isPresent() && !breakpoint.getLogMessage().get().isBlank() ?
+                breakpoint.getLogMessage().get() : "";
+
+        if (logMessage.isEmpty() && condition.isEmpty()) {
+            notifyStopEvent(event);
+            return;
+        }
+
+        // If there's an non-empty user defined log message and and no breakpoint condition, resumes the remote VM
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java:165]:
```suggestion
        // If there's a non-empty user defined log message and no breakpoint condition, resumes the remote VM
```

File: tests/jballerina-debugger-integration-test/src/main/java/org/ballerinalang/debugger/test/utils/DebugOutputListener.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            debugOutputFound = true;
+            debugOutputContext = event;
+            lastOutputLog = event.getOutput();
+            this.cancel();
+        }
+    }
+
+    public TestDAPClientConnector getConnector() {
+        return connector;
+    }
```
Comment (praveennadarajah) [inline tests/jballerina-debugger-integration-test/src/main/java/org/ballerinalang/debugger/test/utils/DebugOutputListener.java:71]: Unused method.

File: tests/jballerina-debugger-integration-test/src/main/java/org/ballerinalang/debugger/test/utils/DebugTestRunner.java [hunk 5]
```diff
@@ -83,7 +84,8 @@
     private static final int MAX_RETRY_COUNT = 3;
     private BMainInstance balClient = null;
     private Process debuggeeProcess;
-    private DebugHitListener listener;
+    private DebugHitListener hitListener;
+    private DebugOutputListener outputListener;
```
Comment (praveennadarajah) [inline tests/jballerina-debugger-integration-test/src/main/java/org/ballerinalang/debugger/test/utils/DebugTestRunner.java:88]: `outputListener` is only used inside `waitForDebugOutput`. In that case, shall make this field to local variable in  `waitForDebugOutput` method?

File: tests/jballerina-debugger-integration-test/src/test/resources/project-based-tests/logpoint-tests/Ballerina.toml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+[package]
+org = "debug_test_resources"
+name = "logpoint_tests"
+version = "0.1.0"
+
+[build-options]
+observabilityIncluded = true
```
Comment (praveennadarajah) [inline tests/jballerina-debugger-integration-test/src/test/resources/project-based-tests/logpoint-tests/Ballerina.toml:7]: Do we need to include this here?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java (hunk 3 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JDIEventProcessor.java]: Fixed with d318b5f

File: tests/jballerina-debugger-integration-test/src/main/java/org/ballerinalang/debugger/test/utils/DebugOutputListener.java (hunk 4 above)
Comment (NipunaRanasinghe) [reply tests/jballerina-debugger-integration-test/src/main/java/org/ballerinalang/debugger/test/utils/DebugOutputListener.java]: Removed with d318b5f

File: tests/jballerina-debugger-integration-test/src/main/java/org/ballerinalang/debugger/test/utils/DebugTestRunner.java (hunk 5 above)
Comment (NipunaRanasinghe) [reply tests/jballerina-debugger-integration-test/src/main/java/org/ballerinalang/debugger/test/utils/DebugTestRunner.java]: Changed in d318b5f

File: tests/jballerina-debugger-integration-test/src/test/resources/project-based-tests/logpoint-tests/Ballerina.toml (hunk 6 above)
Comment (NipunaRanasinghe) [reply tests/jballerina-debugger-integration-test/src/test/resources/project-based-tests/logpoint-tests/Ballerina.toml]: Removed in d318b5f

**Review by praveennadarajah — APPROVED** [review]

**Pull request merged**